controllers/baremetalset_controller.go

```diff
@@ -90,76 +102,139 @@ func (r *BaremetalSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
+	// If BaremetalHosts status map is nil, create it
+	if instance.Status.BaremetalHosts == nil {
+		instance.Status.BaremetalHosts = map[string]ospdirectorv1beta1.BaremetalHostStatus{}
 	}
 
-	// for each available BaremetalHost we Update the reference to use our image
-	for _, availableHost := range availableBaremetalHosts {
-
-		foundBaremetalHost := &metal3v1alpha1.BareMetalHost{}
-		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: availableHost, Namespace: "openshift-machine-api"}, foundBaremetalHost)
-
-		if err != nil {
-			return ctrl.Result{}, err
-		}
+	// Provision / deprovision requested replicas
+	if err := r.ensureBaremetalHosts(instance, provisionServer, sshSecret); err != nil {
+		return ctrl.Result{}, err
+	}
 
-		r.Log.Info(fmt.Sprintf("Updating BaremetalHost: %s", availableHost))
-		foundBaremetalHost.Spec.Online = true
-		foundBaremetalHost.Spec.ConsumerRef = &corev1.ObjectReference{Name: baremetalset.Name, Kind: "BaremetalSet"}
-		foundBaremetalHost.Spec.Image = &metal3v1alpha1.Image{
-			URL:      provisionServer.Status.LocalImageURL,
-			Checksum: fmt.Sprintf("%s.md5sum", provisionServer.Status.LocalImageURL),
-		}
-		err = r.Client.Update(context.TODO(), foundBaremetalHost)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		return ctrl.Result{RequeueAfter: time.Second * 5}, err
+	err = r.Client.Status().Update(context.TODO(), instance)
 
+	if err != nil {
+		r.Log.Error(err, "Failed to update CR status %v")
+		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
 // SetupWithManager - prepare controller for use with operator manager
 func (r *BaremetalSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	openshiftMachineApiSecretsFn := handler.ToRequestsFunc(func(o handler.MapObject) []reconcile.Request {
+		cc := mgr.GetClient()
+		result := []reconcile.Request{}
+		secret := &corev1.Secret{}
+		key := client.ObjectKey{Namespace: o.Meta.GetNamespace(), Name: o.Meta.GetName()}
+		err := cc.Get(context.Background(), key, secret)
+		if err != nil && !errors.IsNotFound(err) {
+			r.Log.Error(err, "Unable to retrieve Secret %v")
+			return nil
+		}
+
+		label := secret.ObjectMeta.GetLabels()
+		// verify object has ownerUIDLabelSelector
+		if uid, ok := label[baremetalset.OwnerUIDLabelSelector]; ok {
+			r.Log.Info(fmt.Sprintf("Secret object %s marked with OSP owner ref: %s", o.Meta.GetName(), uid))
+			// return namespace and Name of CR
+			name := client.ObjectKey{
+				Namespace: label[baremetalset.OwnerNameSpaceLabelSelector],
+				Name:      label[baremetalset.OwnerNameLabelSelector],
+			}
+			result = append(result, reconcile.Request{NamespacedName: name})
+		}
+		if len(result) > 0 {
+			return result
+		}
+		return nil
+	})
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&ospdirectorv1beta1.BaremetalSet{}).
 		Owns(&ospdirectorv1beta1.ProvisionServer{}).
+		Watches(&source.Kind{Type: &corev1.Secret{}},
```

Comment:
what happens when the bmh secret changes? right now I don't see a need to reconcile when a bmh secret changes. We can not update e.g. ssh auth_keys via this controller to access the server. But maybe I miss something

But we might want to reconcile on our metal3v1alpha1.BareMetalHost object and as mentioned above add the power status in our the cr status?

Reply:
This watch might be the wrong approach indeed.  A result of my novice operator knowledge that will hopefully improve. :-)

But yes, watching the `BareMetalHost` object seems proper to add before we finish this draft.

Reply:
also, I think there is a performance hit if you watch objects across namespaces.

Reply:
There indeed may be a performance hit.  I am just replicating the pattern we used for the compute node operator.  This is also what the experts suggested when I asked them about it in one of the operator development meetings in the past few weeks.